# CaSR to Operational Forcing Fields

This notebook converts an integrated `CaSR` forcing field into an operational forcing field suitable for `SalishSeaCast`.

## CaSR information

CaSR file names:

    Integrated_RDPS_P_FB_SFC_2008_2024.nc   Integrated_RDPS_P_PR0_SFC_2008_2024.nc
    Integrated_RDPS_P_FI_SFC_2008_2024.nc   Integrated_RDPS_P_TT_1.5m_2008_2024.nc
    Integrated_RDPS_P_HU_1.5m_2008_2024.nc  Integrated_RDPS_P_UUC_10m_2008_2024.nc
    Integrated_RDPS_P_PN_SFC_2008_2024.nc   Integrated_RDPS_P_VVC_10m_2008_2024.nc

Base path: `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/`

In [1]:
# paths

base_path_casr = '/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/'

fname_FB  = 'Integrated_RDPS_P_FB_SFC_2008_2024.nc'
fname_PRO = 'Integrated_RDPS_P_PR0_SFC_2008_2024.nc'
fname_FI  = 'Integrated_RDPS_P_FI_SFC_2008_2024.nc'
fname_TT  = 'Integrated_RDPS_P_TT_1.5m_2008_2024.nc'
fname_HU  = 'Integrated_RDPS_P_HU_1.5m_2008_2024.nc'
fname_UUC = 'Integrated_RDPS_P_UUC_10m_2008_2024.nc'
fname_PN  = 'Integrated_RDPS_P_PN_SFC_2008_2024.nc'
fname_VVC = 'Integrated_RDPS_P_VVC_10m_2008_2024.nc'

path_FB  = base_path_casr + fname_FB
path_PRO = base_path_casr + fname_PRO
path_FI  = base_path_casr + fname_FI
path_TT  = base_path_casr + fname_TT
path_HU  = base_path_casr + fname_HU
path_UUC = base_path_casr + fname_UUC
path_PN  = base_path_casr + fname_PN
path_VVC = base_path_casr + fname_VVC

In [2]:
import sys
sys.path.append("/home/jqiu/analysis-junqi/Tools-Junqi")

import junqi_nctool as jqnc

### SFC variables

In [3]:
jqnc.nc_disp(path_FB,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_FB_SFC_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_FB_SFC,"(149040, 105, 70)","time, rlat, rlon",Forecast: Downward solar flux at surface,W m**-2
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


In [4]:
jqnc.nc_disp(path_PRO,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_PR0_SFC_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_PR0_SFC,"(149040, 105, 70)","time, rlat, rlon",Forecast: Quantity of precipitation (model) at surface,m
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


In [5]:
jqnc.nc_disp(path_FI,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_FI_SFC_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_FI_SFC,"(149040, 105, 70)","time, rlat, rlon",Forecast: Surface incoming infrared flux at surface,W m**-2
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


In [6]:
jqnc.nc_disp(path_PN,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_PN_SFC_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_PN_SFC,"(149040, 105, 70)","time, rlat, rlon",Forecast: Sea level pressure at surface,mb
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


### 1.5 m variables

In [7]:
jqnc.nc_disp(path_TT,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_TT_1.5m_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_TT_1.5m,"(149040, 105, 70)","time, rlat, rlon",Forecast: Air temperature at 1.5m,deg_C
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


In [8]:
jqnc.nc_disp(path_HU,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_HU_1.5m_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_HU_1.5m,"(149040, 105, 70)","time, rlat, rlon",Forecast: Specific humidity at 1.5m,kg kg**-1
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


### 10 m variables

In [9]:
jqnc.nc_disp(path_UUC,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_UUC_10m_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_UUC_10m,"(149040, 105, 70)","time, rlat, rlon",Forecast: Corrected U-component of the wind (along West-East direction) at 10m,kts
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


In [10]:
jqnc.nc_disp(path_VVC,detailed=True)

### NetCDF Summary

*Path: : `/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/Integrated_RDPS_P_VVC_10m_2008_2024.nc`*

#### Dimensions

Dimension,Size
time,149040
rlat,105
rlon,70


#### Variables

Name,Shape,Dimensions,Long Name,Units
rotated_pole,(),scalar,,
CaSR_v3.2_P_VVC_10m,"(149040, 105, 70)","time, rlat, rlon",Forecast: Corrected V-component of the wind (along South-North direction) at 10m,kts
time,"(149040,)",time,Validity time,
lon,"(105, 70)","rlat, rlon",longitude,degrees_east
lat,"(105, 70)","rlat, rlon",latitude,degrees_north
rlon,"(70,)",rlon,longitude in rotated pole grid,degrees
rlat,"(105,)",rlat,latitude in rotated pole grid,degrees


#### Global Attributes

Attribute,Value
Conventions,CF-1.6
source,https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/
institution,Environment and Climate Change Canada / Environnement et Changement climatique Canada
title,Canadian Surface Reanalysis (CaSR) / Réanalyse Canadienne de Surface (RCaS)
contact,rcas-casr@ec.gc.ca
licence,These data are produced and provided by the Environment and Climate Change Canada. License agreement can be found in https://open.canada.ca/en/open-government-licence-canada or email rcas-casr@ec.gc.ca / Ces données sont produites et fournies par Environnement et Changement climatique Canada. Le contrat de license peut être trouvé dans https://ouvert.canada.ca/fr/licence-du-gouvernement-ouvert-canada ou courriel rcas-casr@ec.gc.ca


### Lon and Lat 

In [ ]:
import xarray as xr
import os

# 拼接完整的文件路径
base_path_casr = '/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/'
fname_FB  = 'Integrated_RDPS_P_FB_SFC_2008_2024.nc'
file_path = os.path.join(base_path_casr, fname_FB)

# 读取 NetCDF 文件
try:
    # 默认引擎通常是 netcdf4或h5netcdf
    ds = xr.open_dataset(file_path)
    
    # 获取 lon 和 lat，并计算它们的最小值和最大值
    # .item() 用于将单个元素的 xarray DataArray 转换为 Python 原生的 float 类型
    lon_min = ds['lon'].min().item()
    lon_max = ds['lon'].max().item()
    
    lat_min = ds['lat'].min().item()
    lat_max = ds['lat'].max().item()
    
    # 打印结果
    print(f"文件: {fname_FB}")
    print("-" * 30)
    print(f"经度 (lon) 范围: {lon_min:.4f}  到  {lon_max:.4f}")
    print(f"纬度 (lat) 范围: {lat_min:.4f}  到  {lat_max:.4f}")
    
    # 关闭数据集释放内存
    ds.close()

except Exception as e:
    print(f"读取文件时出错: {e}")

文件: Integrated_RDPS_P_FB_SFC_2008_2024.nc
------------------------------
经度 (lon) 范围: 229.0648  到  243.7545
纬度 (lat) 范围: 43.3772  到  54.4950


## HRDPS information

In [12]:
path_hrdps='/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc'

jqnc.nc_disp(path_hrdps,detailed=True)

### NetCDF Summary

*Path: : `/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc`*

#### Dimensions

Dimension,Size
time_counter,24
y,266
x,256


#### Variables

Name,Shape,Dimensions,Long Name,Units
atmpres,"(24, 266, 256)","time_counter, y, x",Pressure Reduced to MSL,Pa
nav_lat,"(266, 256)","y, x",latitude,degrees_north
nav_lon,"(266, 256)","y, x",longitude,degrees_east
percentcloud,"(24, 266, 256)","time_counter, y, x",Total Cloud Cover,percent
precip,"(24, 266, 256)","time_counter, y, x",Total Precipitation,kg/m^2
qair,"(24, 266, 256)","time_counter, y, x",Specific Humidity,kg/kg
solar,"(24, 266, 256)","time_counter, y, x",Downward Short-Wave Radiation Flux,W/m^2
tair,"(24, 266, 256)","time_counter, y, x",Temperature,K
therm_rad,"(24, 266, 256)","time_counter, y, x",Downward Long-Wave Rad. Flux,W/m^2
u_wind,"(24, 266, 256)","time_counter, y, x",U-Component of Wind,m/s


#### Global Attributes

Attribute,Value
Conventions,CF-1.0
History,Mon Jan 1 10:36:51 2018: ncks -4 -L4 -O /results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc /results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc created by wgrib2
GRIB2_grid_template,20
NCO,4.4.2


In [5]:
import xarray as xr
import os

# 文件路径

file_path = '/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc'

# 读取 NetCDF 文件
try:
    # 默认引擎通常是 netcdf4或h5netcdf
    ds = xr.open_dataset(file_path)
    
    # 获取 lon 和 lat，并计算它们的最小值和最大值
    # .item() 用于将单个元素的 xarray DataArray 转换为 Python 原生的 float 类型
    lon_min = ds['nav_lon'].min().item()
    lon_max = ds['nav_lon'].max().item()
    
    lat_min = ds['nav_lat'].min().item()
    lat_max = ds['nav_lat'].max().item()
    
    # 打印结果
    print(f"文件: {fname_FB}")
    print("-" * 30)
    print(f"经度 (lon) 范围: {lon_min:.4f}  到  {lon_max:.4f}")
    print(f"纬度 (lat) 范围: {lat_min:.4f}  到  {lat_max:.4f}")
    
    # 关闭数据集释放内存
    ds.close()

except Exception as e:
    print(f"读取文件时出错: {e}")

print(ds.time_counter)

文件: Integrated_RDPS_P_FB_SFC_2008_2024.nc
------------------------------
经度 (lon) 范围: 230.8331  到  240.5296
纬度 (lat) 范围: 45.6649  到  52.1793
<xarray.DataArray 'time_counter' (time_counter: 24)> Size: 192B
array(['2018-01-01T00:00:00.000000000', '2018-01-01T01:00:00.000000000',
       '2018-01-01T02:00:00.000000000', '2018-01-01T03:00:00.000000000',
       '2018-01-01T04:00:00.000000000', '2018-01-01T05:00:00.000000000',
       '2018-01-01T06:00:00.000000000', '2018-01-01T07:00:00.000000000',
       '2018-01-01T08:00:00.000000000', '2018-01-01T09:00:00.000000000',
       '2018-01-01T10:00:00.000000000', '2018-01-01T11:00:00.000000000',
       '2018-01-01T12:00:00.000000000', '2018-01-01T13:00:00.000000000',
       '2018-01-01T14:00:00.000000000', '2018-01-01T15:00:00.000000000',
       '2018-01-01T16:00:00.000000000', '2018-01-01T17:00:00.000000000',
       '2018-01-01T18:00:00.000000000', '2018-01-01T19:00:00.000000000',
       '2018-01-01T20:00:00.000000000', '2018-01-01T21:00:00.0000

### Summary

It seems that `Total Cloud Cover` is missing and I believe that it's not provided by `CaSR`, but that variable is not necessary. See https://hpfx.collab.science.gc.ca/~scar700/rcas-casr/download_CaSR_regions_var_period.html



| CaSR variable description | CaSR units | HRDPS variable name | HRDPS long name | HRDPS units | Notes |
|---|---:|---|---|---:|---|
| Forecast: Corrected V-component of the wind along South-North direction at 10 m | kts | v_wind | V-Component of Wind | m/s  | Same wind component, but units differ: CaSR uses knots, HRDPS uses m/s. |
| Forecast: Corrected U-component of the wind along West-East direction at 10 m | kts | u_wind | U-Component of Wind | m/s  | Same wind component, but units differ: CaSR uses knots, HRDPS uses m/s. |
| Forecast: Specific humidity at 1.5 m | kg kg**-1 | qair | Specific Humidity | kg/kg  | Units are effectively equivalent. |
| Forecast: Air temperature at 1.5 m | deg_C | tair | Temperature | K  | Same temperature variable, but units differ: CaSR uses Celsius, HRDPS uses Kelvin. |
| Forecast: Sea level pressure at surface | mb | atmpres | Pressure Reduced to MSL | Pa  | Same sea-level pressure concept, but units differ: CaSR uses millibars, HRDPS uses pascals. |
| Forecast: Surface incoming infrared flux at surface | W m**-2 | therm_rad | Downward Long-Wave Rad. Flux | W/m^2  | Incoming infrared flux corresponds to downward long-wave radiation. |
| Forecast: Quantity of precipitation model at surface | m | precip | Total Precipitation | kg/m^2  | Same precipitation-related variable, but units differ. For water, 1 kg/m² is approximately 1 mm water equivalent. |
| Forecast: Downward solar flux at surface | W m**-2 | solar | Downward Short-Wave Radiation Flux | W/m^2  | Downward solar flux corresponds to downward short-wave radiation. |
|Not Provided | | percentcloud | Total Cloud Cover | percent | This variable appears in HRDPS but was not included in the provided CaSR variable list.






## Test Conversion and inspection

Try to generate one day of data, and see if there are any mistakes.

In [1]:
import xarray as xr
import numpy as np
import pandas as pd
import os

# 基础路径与配置
base_path = '/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/'
target_date = '2023-03-01' # 先截取 2023 年 3 月 1 日进行测试

# 目标变量名 : (原文件名, 原变量名)
var_mapping = {
    'solar':     ('Integrated_RDPS_P_FB_SFC_2008_2024.nc',  'CaSR_v3.2_P_FB_SFC'),
    'precip':    ('Integrated_RDPS_P_PR0_SFC_2008_2024.nc', 'CaSR_v3.2_P_PR0_SFC'),
    'therm_rad': ('Integrated_RDPS_P_FI_SFC_2008_2024.nc',  'CaSR_v3.2_P_FI_SFC'),
    'tair':      ('Integrated_RDPS_P_TT_1.5m_2008_2024.nc', 'CaSR_v3.2_P_TT_1.5m'),
    'qair':      ('Integrated_RDPS_P_HU_1.5m_2008_2024.nc', 'CaSR_v3.2_P_HU_1.5m'),
    'u_wind':    ('Integrated_RDPS_P_UUC_10m_2008_2024.nc', 'CaSR_v3.2_P_UUC_10m'),
    'atmpres':   ('Integrated_RDPS_P_PN_SFC_2008_2024.nc',  'CaSR_v3.2_P_PN_SFC'),
    'v_wind':    ('Integrated_RDPS_P_VVC_10m_2008_2024.nc', 'CaSR_v3.2_P_VVC_10m')
}

# 初始化目标 Dataset 与坐标
ds_out = xr.Dataset()

# 用 TT(温度) 文件作为参考来获取网格和时间信息
ref_file = os.path.join(base_path, 'Integrated_RDPS_P_TT_1.5m_2008_2024.nc')
ds_ref = xr.open_dataset(ref_file).sel(time=target_date)

y_size, x_size = ds_ref.rlat.size, ds_ref.rlon.size
time_size = ds_ref.time.size  # 获取实际的时间步数量，替代硬编码的 24

# 建立目标格式要求的坐标系
ds_out.coords['y'] = np.arange(y_size)
ds_out.coords['x'] = np.arange(x_size)
# 【修改点 1】直接提取原始数据的标准时间并赋给 time_counter
ds_out.coords['time_counter'] = ds_ref['time'].values 

# 重命名经纬度并赋值 (维度变为 y, x)
ds_out['nav_lat'] = (('y', 'x'), ds_ref['lat'].values)
ds_out['nav_lon'] = (('y', 'x'), ds_ref['lon'].values)
ds_out['nav_lat'].attrs['units'] = 'degrees_north'
ds_out['nav_lon'].attrs['units'] = 'degrees_east'

ds_ref.close()

# 循环读取并转换各变量
print(f"开始处理 {target_date} 的数据...")

for target_var, (fname, source_var) in var_mapping.items():
    fpath = os.path.join(base_path, fname)
    ds_in = xr.open_dataset(fpath).sel(time=target_date)
    
    # 提取数组
    data = ds_in[source_var].values
    
    # 执行单位转换
    if target_var == 'tair':
        data = data + 273.15        # deg_C -> K
        units = 'K'
    elif target_var in ['u_wind', 'v_wind']:
        data = data * 0.514444      # kts -> m/s
        units = 'm/s'
    elif target_var == 'atmpres':
        data = data * 100.0         # mb -> Pa
        units = 'Pa'
    elif target_var == 'precip':
        data = data * 1000.0        # m -> kg/m^2 (等效 1mm)
        units = 'kg/m^2'
    elif target_var in ['solar', 'therm_rad']:
        units = 'W/m^2'
    elif target_var == 'qair':
        units = 'kg/kg'
        
    # 存入新 Dataset，维度固定为 (time_counter, y, x)
    ds_out[target_var] = (('time_counter', 'y', 'x'), data)
    ds_out[target_var].attrs['units'] = units
    
    ds_in.close()

# 补充缺失的云覆盖率变量 (使用 time_size 动态适配时间步长度)
ds_out['percentcloud'] = (('time_counter', 'y', 'x'), np.zeros((time_size, y_size, x_size)))
ds_out['percentcloud'].attrs['units'] = 'percent'

# 强制规范时间变量的编码格式
# dtype 设为 float64 或 int64 防止秒数太大溢出（通常 float64 最稳妥）
out_encoding = {
    'time_counter': {
        'units': 'seconds since 1970-01-01 00:00:00',
        'calendar': 'standard',
        'dtype': 'float64'
    }
}

date_str = pd.to_datetime(target_date).strftime('%Ym%md%d')
output_filename = f'CaSR_y{date_str}.nc'

# 导出时传入 encoding 字典
ds_out.to_netcdf(
    output_filename, 
    unlimited_dims=('time_counter',),
    encoding=out_encoding
)

print(f"生成完毕: {output_filename}")

开始处理 2023-03-01 的数据...
生成完毕: CaSR_y2023m03d01.nc


### Inspection

In [2]:
# Inspection

path_CaSR='/home/jqiu/analysis-junqi/Analysis_Atmospheric_Forcing/Data_CaSR_Conversion/CaSR_y2008m01d01.nc'
path_HRDPS='/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc'


In [5]:
# 随便挑一个原文件读取（
ds_test = xr.open_dataset(ref_file)

# 打印 xarray 解析后的属性 和 隐藏的原始单位
print("【当前解析结果】:\n", ds_test['time'].head(5)) 
print("\n【原始文件中的时间单位】:", ds_test['time'].encoding.get('units', '未找到 units 属性'))

【当前解析结果】:
 <xarray.DataArray 'time' (time: 5)> Size: 40B
array(['2007-12-31T13:00:00.000000000', '2007-12-31T14:00:00.000000000',
       '2007-12-31T15:00:00.000000000', '2007-12-31T16:00:00.000000000',
       '2007-12-31T17:00:00.000000000'], dtype='datetime64[ns]')
Coordinates:
  * time     (time) datetime64[ns] 40B 2007-12-31T13:00:00 ... 2007-12-31T17:...
Attributes:
    standard_name:  time
    long_name:      Validity time
    axis:           T

【原始文件中的时间单位】: hours since 2007-12-31T12:00:00


In [7]:
import netCDF4 as nc
import numpy as np

# Define file paths
path_CaSR = '/home/jqiu/analysis-junqi/Analysis_Atmospheric_Forcing/Data_CaSR_Conversion/CaSR_y2008m01d01.nc'
path_HRDPS = '/results/forcing/atmospheric/GEM2.5/operational/ops_y2018m01d01.nc'

# Target variables for comparison
TARGET_VARS = [
    'time_counter', 'x', 'y', 
    'nav_lat', 'nav_lon', 
    'atmpres', 'percentcloud', 'precip', 
    'qair', 'solar', 'tair', 
    'therm_rad', 'u_wind', 'v_wind'
]

def get_var_info(dataset, var_name):
    """Extract attributes and sample data for a single variable, returning a dictionary."""
    if var_name not in dataset.variables:
        return {"shape": "Missing", "type": "Missing", "units": "Missing", "sample": "Missing"}
    
    var_obj = dataset.variables[var_name]
    info = {
        "shape": str(var_obj.shape),
        "type": str(var_obj.datatype),
        "units": getattr(var_obj, 'units', 'None')  # Display "None" if units attribute is missing
    }
    
    # Safely read a small amount of sample data
    if np.prod(var_obj.shape) > 0:
        try:
            if len(var_obj.shape) == 1:
                sample = var_obj[:3]
            elif len(var_obj.shape) == 2:
                sample = var_obj[0, :3]
            elif len(var_obj.shape) == 3:
                sample = var_obj[0, 0, :3]
            elif len(var_obj.shape) == 4:
                sample = var_obj[0, 0, 0, :3]
            else:
                sample = var_obj[...].flatten()[:3]
            
            # Format array: truncate floats to avoid long strings, take first 3 values
            if sample.dtype.kind == 'f':
                sample_str = "[" + ", ".join([f"{val:.6g}" for val in sample]) + ", ...]"
            else:
                sample_str = "[" + ", ".join([str(val) for val in sample]) + ", ...]"
                
            info["sample"] = sample_str
        except Exception:
            info["sample"] = "Read Error"
    else:
        info["sample"] = "Empty Array"
        
    return info

def compare_nc_files_as_table(file1, file2, vars_to_check):
    print("\nOpening files for side-by-side comparison...")
    try:
        ds1 = nc.Dataset(file1, 'r')
        ds2 = nc.Dataset(file2, 'r')
    except Exception as e:
        print(f"Failed to open files: {e}")
        return
    
    # Set table column widths
    col1_w = 14  # Variable Name
    col2_w = 10  # Attribute
    col3_w = 36  # File 1 Data
    col4_w = 36  # File 2 Data
    
    # Print table header
    sep_line = "-" * (col1_w + col2_w + col3_w + col4_w + 13)
    print(sep_line)
    print(f"| {'Variable':<{col1_w}} | {'Attribute':<{col2_w}} | {'CaSR (File 1)':<{col3_w}} | {'HRDPS (File 2)':<{col4_w}} |")
    print(sep_line)
    
    # Iterate through and compare each variable
    for var in vars_to_check:
        info1 = get_var_info(ds1, var)
        info2 = get_var_info(ds2, var)
        
        # First row: Variable Name and Shape
        print(f"| {var:<{col1_w}} | {'Shape':<{col2_w}} | {info1['shape']:<{col3_w}} | {info2['shape']:<{col4_w}} |")
        # Subsequent rows: Type, Units, Sample (Variable column left blank for clean look)
        print(f"| {'':<{col1_w}} | {'Type':<{col2_w}} | {info1['type']:<{col3_w}} | {info2['type']:<{col4_w}} |")
        print(f"| {'':<{col1_w}} | {'Units':<{col2_w}} | {info1['units']:<{col3_w}} | {info2['units']:<{col4_w}} |")
        print(f"| {'':<{col1_w}} | {'Sample':<{col2_w}} | {info1['sample']:<{col3_w}} | {info2['sample']:<{col4_w}} |")
        print(sep_line)

    ds1.close()
    ds2.close()

# Execute comparison
compare_nc_files_as_table(path_CaSR, path_HRDPS, TARGET_VARS)


Opening files for side-by-side comparison...
-------------------------------------------------------------------------------------------------------------
| Variable       | Attribute  | CaSR (File 1)                        | HRDPS (File 2)                       |
-------------------------------------------------------------------------------------------------------------
| time_counter   | Shape      | (24,)                                | (24,)                                |
|                | Type       | float64                              | float64                              |
|                | Units      | seconds since 1970-01-01             | seconds since 1970-01-01 00:00:00.0 0:00 |
|                | Sample     | [1.19915e+09, 1.19915e+09, 1.19915e+09, ...] | [1.51476e+09, 1.51477e+09, 1.51477e+09, ...] |
-------------------------------------------------------------------------------------------------------------
| x              | Shape      | (70,)                 



| Variable | Attribute | CaSR (File 1) | HRDPS (File 2) |
| --- | --- | --- | --- |
| **time_counter** | Shape | (24,) | (24,) |
|  | Type | float64 | float64 |
|  | Units | seconds since 1970-01-01 | seconds since 1970-01-01 00:00:00.0 0:00 |
|  | Sample | [0, 1, 2, ...] | [1.51476e+09, 1.51477e+09, 1.51477e+09, ...] |
| **x** | Shape | (70,) | (256,) |
|  | Type | int64 | float64 |
|  | Units | None | m |
|  | Sample | [0, 1, 2, ...] | [0, 2500, 5000, ...] |
| **y** | Shape | (105,) | (266,) |
|  | Type | int64 | float64 |
|  | Units | None | m |
|  | Sample | [0, 1, 2, ...] | [0, 2500, 5000, ...] |
| **nav_lat** | Shape | (105, 70) | (266, 256) |
|  | Type | float32 | float64 |
|  | Units | degrees_north | degrees_north |
|  | Sample | [43.3772, 43.4174, 43.4575, ...] | [45.6649, 45.6699, 45.6749, ...] |
| **nav_lon** | Shape | (105, 70) | (266, 256) |
|  | Type | float32 | float64 |
|  | Units | degrees_east | degrees_east |
|  | Sample | [235.872, 235.981, 236.09, ...] | [233.016, 233.044, 233.073, ...] |
| **atmpres** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | Pa | Pa |
|  | Sample | [102580, 102618, 102678, ...] | [101921, 101922, 101924, ...] |
| **percentcloud** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float64 | float32 |
|  | Units | percent | percent |
|  | Sample | [0, 0, 0, ...] | [0, 0, 0, ...] |
| **precip** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | kg/m^2 | kg/m^2 |
|  | Sample | [0, 0, 0, ...] | [0, 0, 0, ...] |
| **qair** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | kg/kg | kg/kg |
|  | Sample | [0.00344162, 0.00322179, 0.00316314, ...] | [0.00547139, 0.00547889, 0.00548639, ...] |
| **solar** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | W/m^2 | W/m^2 |
|  | Sample | [78.8438, 80.5781, 80.4062, ...] | [151.242, 151.242, 151.242, ...] |
| **tair** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | K | K |
|  | Sample | [279.215, 278.611, 277.338, ...] | [283.3, 283.275, 283.25, ...] |
| **therm_rad** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | W/m^2 | W/m^2 |
|  | Sample | [255.942, 254.552, 250.036, ...] | [261.088, 263.311, 263.311, ...] |
| **u_wind** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | m/s | m/s |
|  | Sample | [-2.30252, -2.59893, -2.83304, ...] | [-5.36854, -5.41854, -5.46854, ...] |
| **v_wind** | Shape | (24, 105, 70) | (24, 266, 256) |
|  | Type | float32 | float32 |
|  | Units | m/s | m/s |
|  | Sample | [0.496064, 0.33731, 0.354391, ...] | [-0.907041, -0.877041, -0.837041, ...] |

## Full conversion

Do not run it yet! Not until the `2008-01-01` data are confirmed.

In [ ]:
import xarray as xr
import numpy as np
import pandas as pd
import os
import time
from concurrent.futures import ProcessPoolExecutor, as_completed

def process_single_date(target_date, base_path, output_dir):
    """
    工作进程：处理单一日期的数据。
    (每个进程独立打开文件，避免多进程共享 NetCDF 句柄导致的崩溃)
    """
    date_str_out = target_date.strftime('%Ym%md%d')
    date_query = target_date.strftime('%Y-%m-%d')
    output_filename = os.path.join(output_dir, f'CaSR_y{date_str_out}.nc')

    # 断点续传：如果文件已存在，直接跳过
    if os.path.exists(output_filename):
        return f"跳过 (已存在): {date_query}"

    var_mapping = {
        'solar':     ('Integrated_RDPS_P_FB_SFC_2008_2024.nc',  'CaSR_v3.2_P_FB_SFC'),
        'precip':    ('Integrated_RDPS_P_PR0_SFC_2008_2024.nc', 'CaSR_v3.2_P_PR0_SFC'),
        'therm_rad': ('Integrated_RDPS_P_FI_SFC_2008_2024.nc',  'CaSR_v3.2_P_FI_SFC'),
        'tair':      ('Integrated_RDPS_P_TT_1.5m_2008_2024.nc', 'CaSR_v3.2_P_TT_1.5m'),
        'qair':      ('Integrated_RDPS_P_HU_1.5m_2008_2024.nc', 'CaSR_v3.2_P_HU_1.5m'),
        'u_wind':    ('Integrated_RDPS_P_UUC_10m_2008_2024.nc', 'CaSR_v3.2_P_UUC_10m'),
        'atmpres':   ('Integrated_RDPS_P_PN_SFC_2008_2024.nc',  'CaSR_v3.2_P_PN_SFC'),
        'v_wind':    ('Integrated_RDPS_P_VVC_10m_2008_2024.nc', 'CaSR_v3.2_P_VVC_10m')
    }

    try:
        ds_out = xr.Dataset()

        # 1. 提取网格基准和时间信息 (用当天的数据提取，耗时极短)
        ref_file = os.path.join(base_path, 'Integrated_RDPS_P_TT_1.5m_2008_2024.nc')
        with xr.open_dataset(ref_file) as ds_ref_full:
            ds_ref = ds_ref_full.sel(time=date_query)
            y_size, x_size = ds_ref.rlat.size, ds_ref.rlon.size
            lat_matrix = ds_ref['lat'].values
            lon_matrix = ds_ref['lon'].values
            time_values = ds_ref['time'].values  # 提取真实时间
            time_size = ds_ref.time.size         # 动态获取时间步数

        ds_out.coords['y'] = np.arange(y_size)
        ds_out.coords['x'] = np.arange(x_size)
        ds_out.coords['time_counter'] = time_values # 赋真实 datetime64 值
        
        ds_out['nav_lat'] = (('y', 'x'), lat_matrix)
        ds_out['nav_lon'] = (('y', 'x'), lon_matrix)
        ds_out['nav_lat'].attrs['units'] = 'degrees_north'
        ds_out['nav_lon'].attrs['units'] = 'degrees_east'

        # 2. 循环处理各变量
        for target_var, (fname, source_var) in var_mapping.items():
            fpath = os.path.join(base_path, fname)
            
            # 使用 with 语句确保读取后立刻释放文件锁
            with xr.open_dataset(fpath) as ds_in_full:
                day_slice = ds_in_full[source_var].sel(time=date_query).values

            # 物理单位转换
            if target_var == 'tair':
                data, units = day_slice + 273.15, 'K'
            elif target_var in ['u_wind', 'v_wind']:
                data, units = day_slice * 0.514444, 'm/s'
            elif target_var == 'atmpres':
                data, units = day_slice * 100.0, 'Pa'
            elif target_var == 'precip':
                data, units = day_slice * 1000.0, 'kg/m^2'
            elif target_var in ['solar', 'therm_rad']:
                data, units = day_slice, 'W/m^2'
            elif target_var == 'qair':
                data, units = day_slice, 'kg/kg'
                
            ds_out[target_var] = (('time_counter', 'y', 'x'), data)
            ds_out[target_var].attrs['units'] = units

        # 3. 添加云覆盖 (动态适配时间维大小)
        ds_out['percentcloud'] = (('time_counter', 'y', 'x'), np.zeros((time_size, y_size, x_size), dtype=np.float32))
        ds_out['percentcloud'].attrs['units'] = 'percent'

        # 4. 导出 (添加 unlimited_dims 和 encoding)
        out_encoding = {
            'time_counter': {
                'units': 'seconds since 1970-01-01 00:00:00',
                'calendar': 'standard',
                'dtype': 'float64'
            }
        }
        
        ds_out.to_netcdf(
            output_filename, 
            format='NETCDF4', 
            unlimited_dims=('time_counter',), 
            encoding=out_encoding
        )
        return f"成功生成: {date_query}"

    except Exception as e:
        return f"失败 {date_query} 报错: {str(e)}"

def run_parallel_conversion(base_path, output_dir, start_date='2008-01-01', end_date='2024-12-31', max_workers=8):
    os.makedirs(output_dir, exist_ok=True)
    date_range = pd.date_range(start=start_date, end=end_date, freq='D')
    total_days = len(date_range)
    
    print(f"开始并行处理 {total_days} 天的数据，使用 {max_workers} 个进程...")
    start_time = time.time()

    success_count = 0
    # 使用进程池
    with ProcessPoolExecutor(max_workers=max_workers) as executor:
        # 提交所有任务到进程池
        futures = {
            executor.submit(process_single_date, date, base_path, output_dir): date 
            for date in date_range
        }
        
        # 实时获取完成状态
        for i, future in enumerate(as_completed(futures), 1):
            result = future.result()
            
            if "成功" in result or "跳过" in result:
                success_count += 1
            
            # 每完成 50 个打印一次进度，防止日志刷屏
            if i % 50 == 0 or i == total_days:
                elapsed = time.time() - start_time
                speed = i / elapsed
                print(f"[{i}/{total_days}] {result} (平均速度: {speed:.2f} 天/秒)")

    total_time = time.time() - start_time
    print(f"\n全部完成！耗时: {total_time/60:.2f} 分钟。成功处理: {success_count}/{total_days} 天。")

if __name__ == "__main__":
    SOURCE_PATH = '/ocean/jqiu/Atmospheric_RDPS/2008_2024_Integrated/'
    OUTPUT_PATH = '/ocean/jqiu/CaSR/'
    
    # max_workers 建议设置为 CPU 核心数的 50%~80%
    run_parallel_conversion(
        base_path=SOURCE_PATH, 
        output_dir=OUTPUT_PATH,
        start_date='2008-01-01', 
        end_date='2024-12-31',
        max_workers=6  # <--- 这里可以根据服务器性能调整
    )